File: server/src/models/family.py
"""Family model definitions for the genealogy application."""

from datetime import date as date_type
from typing import List, Optional, TYPE_CHECKING
from uuid import UUID, uuid4

from sqlmodel import Field, Relationship, SQLModel

from ..constants import PERSONS_TABLE_ID

if TYPE_CHECKING:
    from .person import Person
    from .event import Event
    from .child import Child


class FamilyBase(SQLModel):
    """Base Family model with common fields."""

    husband_id: Optional[UUID] = Field(default=None, foreign_key=PERSONS_TABLE_ID)
    wife_id: Optional[UUID] = Field(default=None, foreign_key=PERSONS_TABLE_ID)
    marriage_date: Optional[date_type] = Field(default=None)
    marriage_place: Optional[str] = Field(default=None, max_length=200)
    notes: Optional[str] = Field(default=None)


class Family(FamilyBase, table=True):
    """Family model for database storage."""

    __tablename__ = "families"

    id: UUID = Field(default_factory=uuid4, primary_key=True)

    husband: Optional["Person"] = Relationship(
        back_populates="families_as_husband",
        sa_relationship_kwargs={"foreign_keys": "[Family.husband_id]"},
    )
    wife: Optional["Person"] = Relationship(
        back_populates="families_as_wife",
        sa_relationship_kwargs={"foreign_keys": "[Family.wife_id]"},
    )
    events: List["Event"] = Relationship(
        back_populates="family",
        sa_relationship_kwargs={"cascade": "all, delete-orphan"},
    )
    children: List["Child"] = Relationship(
        back_populates="family",
        sa_relationship_kwargs={"cascade": "all, delete-orphan"},
    )


class FamilyCreate(FamilyBase):
    """Family model for creation requests."""


class FamilyRead(FamilyBase):
    """Family model for read responses."""

    id: UUID


class FamilyUpdate(SQLModel):
    """Family model for update requests."""

    husband_id: Optional[UUID] = Field(default=None, foreign_key=PERSONS_TABLE_ID)
    wife_id: Optional[UUID] = Field(default=None, foreign_key=PERSONS_TABLE_ID)
    marriage_date: Optional[date_type] = Field(default=None)
    marriage_place: Optional[str] = Field(default=None, max_length=200)
    notes: Optional[str] = Field(default=None)


class FamilySearchResult(SQLModel):
    """Family model for search results."""

    id: UUID
    husband_name: Optional[str] = None
    wife_name: Optional[str] = None
    husband_sex: Optional[str] = None  # M, F, or U
    wife_sex: Optional[str] = None  # M, F, or U
    marriage_date: Optional[date_type] = None
    marriage_place: Optional[str] = None
    children_count: int = 0
    summary: str  # Human-readable summary like "John Doe & Jane Smith (1920)"


class FamilyDetailResult(FamilyRead):
    """Family model for detailed responses with related data."""

    husband: Optional[dict] = None
    wife: Optional[dict] = None
    children: List[dict] = []
    events: List[dict] = []

File: client/src/types/family.ts
export interface Person {
  id: string
  first_name: string
  last_name: string
  sex: 'M' | 'F' | 'U'
  birth_date?: string
  death_date?: string
  birth_place?: string
  death_place?: string
  occupation?: string
  notes?: string
  has_own_family?: boolean
  own_families?: OwnFamily[]
}

export interface OwnFamily {
  id: string
  marriage_date?: string
  marriage_place?: string
  spouse?: {
    id: string
    name: string
    sex: 'M' | 'F'
  }
  events?: Event[]
}

export interface Child {
  id: string
  family_id: string
  person_id: string
  person?: Person
}

export interface Event {
  id: string
  family_id: string
  type: string
  date?: string
  place?: string
  description?: string
}

export interface FamilySearchResult {
  id: string
  husband_name?: string
  wife_name?: string
  marriage_date?: string
  marriage_place?: string
  children_count: number
  summary: string
}

export interface FamilyDetailResult {
  id: string
  husband_id?: string
  wife_id?: string
  marriage_date?: string
  marriage_place?: string
  notes?: string
  husband?: Person
  wife?: Person
  children: Child[]
  events: Event[]
}

export interface FamilySearchParams {
  q?: string
  family_id?: string
  limit?: number
}

export interface UploadResult {
  message: string
  persons_created: number
  families_created: number
  events_created: number
  children_created: number
}

export interface FamilyRead {
  id: string
  husband_id?: string
  wife_id?: string
  marriage_date?: string
  marriage_place?: string
  notes?: string
}

export interface FamilyManagementParams {
  skip?: number
  limit?: number
}

export interface FamilyDetail {
  id: string
  husband_id?: string
  wife_id?: string
  marriage_date?: string
  marriage_place?: string
  notes?: string
  husband?: Person
  wife?: Person
  children?: Child[]
  events?: Event[]
}
